common/main/java/com/couchbase/lite/internal/replicator/AbstractCBLWebSocket.java

```diff
@@ -326,9 +368,14 @@ protected void requestClose(int status, String message) {
             return;
         }
 
-        if (closing.getAndSet(true)) {
-            Log.v(TAG, "CBLWebSocket already closing.");
-            return;
+        switch (state.getAndSet(State.REQUEST_CLOSED)) {
```

Comment:
This is the main modification against couchbase/couchbase-lite-java-ce-root#13

In this method, `requestClose`, it used a `webSocket == null` to tell if the webSocket handshake was successful, because the `webSocket` field would not be set until calling `onOpen` method.

But we need more than that, on a situation that the core decide to close the webSocket before handshake is done. Then, okhttp should be notified, and have the resources released. Thus, we need to access the `webSocket` field here to call `WebSocekt.cancel`. After calling cancel, okhttp reacts to it, and call `onFailure` later, telling `AbstractCBLWebSocket` and core the failure, so that replication fails (and can be restarted from outside).

But in `requestClose`, how do we know if the handshake is done or not? The old way (checking `webSocket == null`) does not work any longer. So a `State` class is there to hold this information. The old `closing` AtomicBoolean is also covered by the `State` and is removed.

As `onOpen` and `requestClose` is possibly be called in different thread at the same time (consider this situation: the websocket handshake response is received, at the same moment core decide to close it), several code here is for handling the race condition. Please check if something got missed.

Reply:
perfect.  Thanks.  On it.